### PR TITLE
Fixed #25544 -- Removed duplicates in prefetch_related

### PR DESCRIPTION
**Pull request by Ian-Foote**

### CODE REVIEW

File: tests/prefetch_related/tests.py [hunk 1]
```diff
@@ -223,6 +224,27 @@ def test_invalid_final_lookup(self):
...
+
+        sql = queries[-1]['sql']
+        self.assertEqual(sql.count(self.author1.name), 1)
+
+    def test_m2m_then_reverse_one_to_one_object_ids(self):
+        with CaptureQueriesContext(connection) as queries:
+            list(Book.objects.prefetch_related('authors__authorwithage'))
+
+        sql = queries[-1]['sql']
+        self.assertEqual(sql.count(str(self.author1.id)), 1, sql)
```
Comment (Ian-Foote) [inline tests/prefetch_related/tests.py]: I'm missing a case here for [`ForwardManyToOneDescriptor`](https://github.com/django/django/blob/c3531d2f208c5a039a190ec36a7349f627f13f01/django/db/models/fields/related_descriptors.py#L114-L143) which already [partially fixes](https://github.com/django/django/blob/c3531d2f208c5a039a190ec36a7349f627f13f01/django/db/models/fields/related_descriptors.py#L131) this issue.

**Comment by carljm** [comment]
This looks fine, but it makes me wonder whether it ought to be fixed more generally deeper in the ORM, inside the implementation of `__in` lookups? Anytime the ORM receives duplicate values in an `__in` lookup (even if they are given explicitly), there is never any value in pushing those duplicates to the database, is there? Did you consider this option in the implementation?

**Comment by Ian-Foote** [comment]
I will investigate that possibility.

**Comment by Ian-Foote** [comment]
That works, and is cleaner than my previous implementation.

**Comment by charettes** [comment]
The test failure might require writing a special `ArrayIn` lookups that overrides its `get_prep_lookup()` method to return a list of tuples to make it hashable.

**Comment by Ian-Foote** [comment]
I've added the suggested `ArrayIn` lookup.

**Comment by carljm** [comment]
I'm not sure how to interpret the Windows build failure here, but it looks more like an infrastructure failure than a problem with this PR. I don't think there's anything OS-specific here. Marking the ticket RFC. Thanks for the work on this, @Ian-Foote.

(Ping @mjtamlyn in case you haven't seen this and have any additional information we should consider here.)

**Comment by carljm** [comment]
Oops, it seems this now has conflicts with latest master and needs rebasing.

And in further thinking about it, I realized that doing it this way is also potentially backwards-incompatible for any other third-party field type that can accept non-hashable values in an `__in` lookup (like, for instance, a third-party implementation of an ArrayField).

So maybe @akaariai has the best option -- just wrap the attempt to de-duplicate in a `try: ... except TypeError:` and use the originally given list if conversion to a set fails.

**Comment by charettes** [comment]
Let's stick to Anssi's approach then.

**Comment by Ian-Foote** [comment]
I've added the `try: ... except TypeError:`.

**Comment by mjtamlyn** [comment]
Given we have also fixed a wider issue here, I'd suggest a test for that would be in order.

e.g `Foo.objects.filter(name__in=['Ian', 'Carl', 'Simon', 'Ian'])`

**Comment by charettes** [comment]
A test with a generator of unhashable objects would also be required.

I think we'll need to first cast to a `list` and then try to cast to a `set` to make sure the generator is not exhausted.

**Comment by Ian-Foote** [comment]
I added a test for using `In` on a `HStoreField` with a generator, but it appears that `self.rhs` is already a `list` whenever it enters `In.process_rhs`. I'm not clear whether this is guaranteed or something I should not be relying on.

**Comment by charettes** [comment]
@Ian-Foote is it also the case when using the `in` lookup with a `HStoreField`?

**Comment by Ian-Foote** [comment]
@charettes That's covered by the new test. Did you mean `JSONField`?

**Comment by charettes** [comment]
Sorry I edited my comment, I meant "a field different from an `HStoreField`". Unsure how this could be tested on the other end. I guess a `Model` subclass with an `__hash__` method raising an exception could do.

**Comment by charettes** [comment]
@Ian-Foote, just to make sure I was clear about my previous comment. I had the following in mind:

```python
class UnhashableTestModel(TestModel):
    class Meta:
        proxy = True

    def __hash__(self):
        raise Exception

def test_in_unshable_generator(self):
    test_model = UnhashableTestModel.objects.create()
    ModelReferingTestModel.objects.create(test_model=test_model)

    def generator():
      yield test_model

    ModelReferingTestModel.objects.filter(test_model__in=generator())
```

I just want to make sure the initial `set()` won't exhaust the generator and the assumption you made about the `HStoreField` is also true for foreign keys.

**Comment by Ian-Foote** [comment]
@charettes I've tried to implement your test, but `self.rhs` in the `In` lookup becomes a list of integers - I assume primary keys. This means the `__hash__` isn't triggered by the `set`.

**Comment by charettes** [comment]
@Ian-Foote alright, I guess this might not be required then!

**Comment by timgraham** [comment]
merged in 86eccdc8b67728d84440a46e5bf62c78f2eddf6d, thanks!

**Pull request closed without merging**